### DISCUSSION TITLE
parsing native Enum arrays using Zod and slonik

### DISCUSSION THREAD

**Discussion by anmolint** [gajus/slonik#502] [Q&A]
Hi , I have been working on sloink for my personal project and found it quite helpful  but  I  have a question 
I have a table defined as 
```
(
    userId        uuid primary key default uuid_generate_v4() not null,
    email          text                                   not null,
    mobile         text,
    name           text,
    alertChannel alert_methold [] DEFAULT '{"email"}',
    created_at     timestamp with time zone default now() not null,
    updated_at     timestamp with time zone
);
```
with alert_channel defined as  
```
`create type alert_methold as enum ('email',  'push_notifications');`
```
when running a fetch query 
the client errors out telling it did not receive the correct data
the zod schema is defined as
```
`export const UserRecord = z.object({
  user_id: z.string().uuid(), 
  email: z.string().email(),
  mobile: z.string().optional(),
  name: z.string().optional(),
  alertChannel : z.array(z.nativeEnum(AlertChannel)).optional(), 
  created_at: z.instanceof(Date).nullable().optional(),
  updated_at: z.instanceof(Date).nullable().optional(),
});`
```
where enum is defined as
```
enum AlertChannel {
  Email = 'email',
  PushNotifications = 'push_notifications',
}
```
the query
```
export function selectUserById(
    conn: DatabaseTransactionConnection,
    userId: string
): Promise<UserRecord | null> {
    return conn.maybeOne(sql.type(UserRecord)`
        select user_id,
               email,
               mobile,
               name,
               alert_channel,
               created_at,
               updated_at,
               deleted_at
        from ecommerce.users
        where user_id = ${userId};
    `);
}
```
but this results in error from alert_channel any help appreciated thanks  😄

**Comment by nathanchapman** [comment]
Hey @anmolint!

Slonik properly emits an array for built-in types, but for any custom enum array, values are returned as `"{VALUE,VALUE}"` (string) rather than `["VALUE", "VALUE"]` (array of strings).

This is where the [Type Parsers](https://github.com/gajus/slonik/tree/0fc4c2a23515fa34da75f176d0820159fba1df82#type-parsers) come into play.

You can pass these in the `typeParsers` field of `ClientConfiguration` when calling `createPool`:
https://github.com/gajus/slonik/tree/0fc4c2a23515fa34da75f176d0820159fba1df82#api

You can simply set your type parser to:

```
{
    name: 'alert_methold',
    parse: (value) => value,
},
```

I wish this wasn't necessary, but unfortunately it is 😅

**Comment by firxworx** [comment]
I would suggest looking into const assertions everywhere you are using TS enums as the latter are fraught with gotchas.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions

With zod you then use `z.enum()` vs. `z.nativeEnum()`. I have found better results overall and don't recall issues with them off-hand with slonik in similar cases (though its been a while).
